### ISSUE TITLE
Feature Request: Guard against the CMake Binary Directory being the same as the CMake Source Directory

### ISSUE THREAD

**Issue by lumberyard-employee-dm** [o3de/o3de#7977]
**Is your feature request related to a problem? Please describe.**
In CMake the default binary directory if not specified is in the current working directory.
Now this can easily temporary build files to be dumped into the directory of the engine or a project if the user doesn't specify a binary directory during CMake configuration.
If the user current working directory is the <engine-root>, running the `cmake` command without any arguments will pollute the <engine-root> directory with the CMakeCache.txt, as well as other build specific files

**Describe the solution you'd like**
Add a check to guard against having the CMake binary directory be the exact same directory as the cmake source directory.
This isn't meant prevent "in-source" builds, as users can still specify `cmake -B <engine-root>/build/windows` as a build directory, but it would help prevent a user running `cmake` with no arguments having their engine directory polluted.
This guard can be put inside of the root [CMakeLists.txt](../blob/development/CMakeLists.txt)
```
if (CMAKE_SOURCE_DIR STREQUAL CMAKE_BINARY_DIR)
    message(FATAL_ERROR "CMake binary directory cannot be the same as the source. Please specify a build directory to \"cmake -B <build-dir>\" when configuring.")
endif()
```

**Describe alternatives you've considered**
Another alternative is enforce the CMake binary directory that is "in-source" to only be within a directory named "build" instead.
In this case the "build" directory would be an allowlist for where the build directory can be stored.
```
cmake_path(IS_PREFIX CMAKE_SOURCE_DIR "${CMAKE_BINARY_DIR}" in_source_binary_dir)
cmake_path(SET allowed_build_dir_path "${CMAKE_SOURCE_DIR}/build")
cmake_path(IS_PREFIX allowed_build_dir_path "${CMAKE_BINARY_DIR}" build_dir_allowed)
if (in_source_binary_dir AND NOT build_dir_allowed)
    message(FATAL_ERROR "CMake binary directory that is in-source, must be under the \"${allowed_build_dir_path}\" path.")
endif()
```

**Additional context**

**Comment by amzn-changml** [comment]
Yes, this seems like something we should take on. Accepting this and putting into backlog.
